feat(tofu): add ESM support

boneskull opened the pull request:
This makes `tofu` ESM-aware.

@kumavis originally implemented most of this.  I changed it such that:

- It uses the same granularity as the CJS import detection (module level)
- It considers `export ... from` syntax (re-export)

boneskull committed:
feat(tofu): add ESM support

This makes `tofu` ESM-aware.

legobeat commented on packages/tofu/src/inspectSource.js:
```diff
@@ -133,44 +133,38 @@ function inspectGlobals(
+    if (
+      importSource &&
+      ((packagesToInspect && pkgsToInspect.has(importSource)) ||
+        !packagesToInspect)
```
Nit: I feel the added parentheses in the second clause don't help and this rewrite makes it more obvious. WDYT?
```suggestion
      !packagesToInspect
      || pkgsToInspect.has(importSource) && importSource
```

boneskull replied:
prettier does this. it does it because not everyone always remembers order of operations

naugtur replied:
This is confusing to me for other reasons. Since we're turning packagesToInspect into the pkgsToInspect set, could we limit ourselves to using just one? I'd prefer if packagesToInspect were not uttered below the first line of the function. 

And since the empty set is not falsy, but doesn't contain anything, `packagesToInspect && pkgsToInspect.has(importSource)) ||       !packagesToInspect` becomes `pkgsToInspect.has(importSource)) || pkgsToInspect.size===0`. Amirite?

boneskull committed:
chore(tofu): deprecate inspectImports in lieu of inspectRequires

`inspectImports` is a misnomer, because it's actually inspecting CJS requires.  To mitigate any confusion, `inspectRequires` is recommended.  `inspectRequires` is aliased as `inspectImports` and deprecated.

The `@deprecate` tag will show usages of the deprecated symbol in ~~strikethrough~~ in VS Code.

leotm reviewed (commented):
how do we feel about normalising the naming to

```
  inspectRequires,
  inspectImports,
  inspectDynamicRequires,
  inspectDynamicImports, // in future
```

or

```
  inspectCjsRequires,
  inspectEsmImports,
  inspectDynamicCjsRequires,
  inspectDynamicEsmImports, // in future
```

(and presuming dynamic esm imports out of scope for now)

boneskull committed:
chore(tofu): simplify return value of inspectEsmImports

boneskull committed:
fix(tofu): fix type errors

naugtur commented on packages/tofu/src/findGlobals.js:
```diff
@@ -3,26 +3,36 @@
-  'CatchClause',
-  'ArrayPatten',
+  'MetaProperty',
   'RestElement',
```
I hope that's just a reorder. Don't have the patience to verify :facepalm:

boneskull replied:
yes, but `MetaProperty` is new, and refers to `import.meta`

naugtur commented on packages/tofu/src/inspectSource.js:
```diff
@@ -175,60 +172,55 @@ function inspectGlobals(
+    ExportNamedDeclaration: (path) => {
+      handleNodePath(path)
     },
+    ImportDeclaration: handleNodePath,
```
Why is it wrapped in one case and not the other?

naugtur commented on packages/tofu/src/inspectSource.js:
```diff
@@ -175,60 +172,55 @@ function inspectGlobals(
-        return
-      }
-      const importSource = source.value
-      specifiers.forEach((spec) => {
```
How is all this no longer needed? I'm gonna need to dig into this some more. For now what I'm seeing here is we add stuff to imports regardless of all the specifiers processing that was happening here and the case where importName is modified is no longer handled. Not sure why.

boneskull replied:
`ImportDeclaration` handles everything we care about

naugtur replied:
You sound very certain. I have no reason to disbelieve that until we see stuff missing from policies.

naugtur commented on packages/tofu/test/inspectEsmImports.spec.js:
```diff
@@ -29,58 +26,40 @@ testInspect(
+    'package12',
+    'package13',
+    'package14',
+  ]
```
This doesn't seem equivalent. Are we dropping information that's actually unnecessary?

boneskull replied:
It's not equivalent with the previous behavior.  We're dropping the granularity down to module level.  Previously it was granular to named exports.  This is why e.g., `package06/path/to/specific/un-exported/file` only appears once.

naugtur replied:
sounds like a big improvement.

boneskull committed:
chore(tofu): peer review changes
